Add component for binary sensor groups

brianegge opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
This change allows you to create a group of binary sensors with a device class, where the group itself is a binary sensor with a configurable device class.

### Background
If you create a group of binary sensors they do not have the device class of the sensors they contain. Workaround are to create a template.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
If my code was more clever, it would do this anytime you have a heterogeneous group of sensors. But it is not, so in the yaml you must define the device class and you should be sure all the sensors are of that class - though not strictly required. 

 * Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/19297

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository] https://github.com/home-assistant/home-assistant.io/pull/19297

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

brianegge committed:
Accidental push over prior commit

brianegge committed:
Add component for binary sensor groups
https://github.com/home-assistant/home-assistant.io/pull/19239

emontnemery commented on homeassistant/components/group/binary_sensor.py:
```diff
@@ -0,0 +1,130 @@
+    {
+        vol.Required(CONF_ENTITIES): cv.entities_domain(BINARY_SENSOR_DOMAIN),
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_UNIQUE_ID): cv.string,
```
`unique_id` is normally not user configurable.
If we want to support it for groups, this should preferably be moved to GroupEntity though (in a follow-up PR).

brianegge replied:
I copied this from light,cover, and media player in the same package.

emontnemery commented on homeassistant/components/group/binary_sensor.py:
```diff
@@ -0,0 +1,130 @@
+        self._attr_unique_id = unique_id
+        self._device_class = device_class
+        self._state: str | None = None
+        self.mode = any
+        if mode:
+            self.mode = all
```
This is already done in `GroupEntity`

brianegge replied:
OK, I was trying to call GroupEntity's __init__ method but flake8 kept complaining. Maybe the super is getting called, but it seemed this method was needed after I removed the explicit call to init.

emontnemery replied:
My mistake, I confused `GroupEntity` and `Group`, sorry about that.

emontnemery commented on tests/components/group/test_binary_sensor.py:
```diff
@@ -0,0 +1,94 @@
+from homeassistant.setup import async_setup_component
+
+
+async def test_default_state(hass):
```
This test, or a new one, should verify that the configurable unique_id and attributes (name, device class, icon if added) are working. The same improvement should be done for the other groups in a follow-up.

brianegge replied:
Aren't the tests for the base classes already covering this?

emontnemery replied:
I don't think so, since the base class don't support configuring `unique_id` etc.?

brianegge replied:
I added a test for it. They can later be removed from all classes in the package if the attribute is not recommended.

emontnemery commented on tests/components/group/test_binary_sensor.py:
```diff
@@ -0,0 +1,94 @@
+    assert entry.unique_id == "unique_identifier"
+
+
+async def test_state_reporting(hass):
```
this tests the behavior with `all = True`, we should add one which tests the behavior with `all = False` too. The same improvement should be done for the other groups in a follow-up.

brianegge replied:
Agreed. The odd thing about using the binary sensor base is the state is forced to be either true/false, and the unavailable state is removed, so you can't actually return unavailable if the member entities are unavailable. Anyway, happy to add the addition test case or cases if this PR has any hope of being merged.

emontnemery replied:
Availability is still supported, but it's done by implementing the availability property attribute, not by writing to the state directly.

Either:
```
@property
    def available(self) -> bool:
        """Return True if entity is available."""
        return ...
```

Or:
```py
self._attr_available = ...
```

brianegge commented on homeassistant/components/group/binary_sensor.py:
```diff
@@ -0,0 +1,130 @@
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_ENTITIES): cv.entities_domain(BINARY_SENSOR_DOMAIN),
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
```
My first attempt was to place this in the `group` config schema, which turns out to be worthless. The reason the UI and everything else then doesn't treat the device like a binary sensor. With this method, if I create a group of leak sensors the UI is going to show wet/dry and the correct icon. In the group schema you will get the default three dot icon and a true or false text.

emontnemery replied:
That's OK, this can be improved in a separate PR, also refactoring the other groups.

brianegge commented on homeassistant/components/group/binary_sensor.py:
```diff
@@ -0,0 +1,130 @@
+            await self.async_update()
+            return
+
+        await super().async_added_to_hass()
```
I don't really understand this code or the suggestion. I copied this from light and cover in the same package. It seems it's better to be consistent, and if the package is refactored, all of the classes could be fixed at the same time.

emontnemery replied:
OK, let's do the refactoring in a separate PR 👍

brianegge committed:
Add test for any case

emontnemery reviewed (commented):
This is good to go once:
 - The tests have been improved to ensure all configuration variables work as intended.
 - availability is implemented in the same way as for other groups, it should work to just do:
    ```py
   self._attr_available = any(state.state != STATE_UNAVAILABLE for state in states)
```

brianegge committed:
Add unavailable attribute and tests for unique_id

brianegge commented:
Added the `_attr_available`. I added a test for unique_id which was the only input not tested. 

```
        vol.Required(CONF_ENTITIES): cv.entities_domain(BINARY_SENSOR_DOMAIN),
        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
        vol.Optional(CONF_UNIQUE_ID): cv.string,
        vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
        vol.Optional(CONF_ALL): cv.boolean,
```

brianegge committed:
Added tests for attributes

link to documentation: https://github.com/home-assistant/home-assistant.io/pull/19297

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#55365).

Once you do that we will be able to review and accept this pull request.

Thanks!

brianegge commented:
Not sure how to get the docs-missing test to pass. I've added the link in the PR and description. https://github.com/home-assistant/home-assistant.io/pull/19297

emontnemery reviewed (approved):
Thanks @brianegge 👍

emontnemery commented:
The docs missing label needs to be manually cleared.

Thanks for finishing off the tests! 👍

MartinHjelmare commented on homeassistant/components/group/binary_sensor.py:
```diff
@@ -0,0 +1,133 @@
+        if STATE_UNAVAILABLE in filtered_states:
+            self._attr_is_on = None
+        else:
+            states = list(map(lambda x: x == STATE_ON, filtered_states))
```
We prefer using a list comprehension instead of map + lambda.

MartinHjelmare commented on homeassistant/components/group/binary_sensor.py:
```diff
@@ -0,0 +1,133 @@
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_UNIQUE_ID): cv.string,
+        vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+        vol.Optional(CONF_ALL): cv.boolean,
```
We don't have an all option for the other group platforms. Eg see the light platform which also has a binary state. I think we should be really careful before we add new options to these platforms.

emontnemery replied:
I think these platforms should be refactored so the features are more similar and also to deduplicate common code.
